### PR TITLE
drivers: i2c: sam0: Continue write/read if next message allows it

### LINKED ISSUE #36857: i2c_samd0.c burst write not compatible with ssd1306.c
Running Zephyr with a SSD1306 compatible OLED screen does not work over I2C

I am running with a [seeeduino xiao](https://wiki.seeedstudio.com/Seeeduino-XIAO/) and its accompanying [dev kit](https://wiki.seeedstudio.com/Seeeduino-XIAO-Expansion-Board/) which features a SSD1306 compatible 128x64 pixels OLED screen, accessed over the I2C bus.

Comparing with a similar library from the Arduino ecosystem I saw that the I2C packets of Zephyr was incorrectly chopped up into smaller chunks. I tracked it down and it seems that it comes from the usage of `i2c_burst_write()`, but I don't know what is the supposed behavior of the I2C driver. Is it okay for it to inject STOP conditions between each I2C message? There is a warning in the header file which seems to indicate that it is.

**To Reproduce**
1. checkout the below zephyr version
2. apply the changes in the below attached patch (except the one for the ssd1306.c file which contains a workaround)
3. unpack blinken.zip below
4. `west build -b seeeduino_xiao`
5. `west flash`
6. See error

**Expected behavior**
A stripey pattern on the display

**Impact**
Showstopper for any application using this driver combo (and possibly others with the ssd1306?)

**Logs and console output**
Expected comms:
image

Actual comms:
image

**Environment (please complete the following information):**
 - Manjaro Linux
 - zephyr-v2.6.0-740-gc32a96af6ade

**Additional context**
[0001-seeduino-xiao-add-i2c-support-fix-ssd1306-burst-writ.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/6794058/0001-seeduino-xiao-add-i2c-support-fix-ssd1306-burst-writ.patch.txt)
[blinken.zip](https://github.com/zephyrproject-rtos/zephyr/files/6794100/blinken.zip)

### PR DESCRIPTION
**Pull request by rustypig91**
Update sam0 i2c driver to directly send/receive next message if it is
in the same direction and the current message has no stop or restart
flags. Seems like in some drivers this is the expected behaviour.

Fixes #36857

Signed-off-by: Christoffer Zakrisson <rustypig91@gmail.com>

### CODE REVIEW

File: drivers/i2c/i2c_sam0.c [hunk 1]
```diff
@@ -137,6 +139,17 @@ static void i2c_sam0_isr(const struct device *dev)
 		return;
 	}
 
+	/*
+	 * Directly send/receive next message if it is in the same direction and
+	 * the current message has no stop or restart flags.
+	 */
+	const int continue_next = (data->msg.size == 1) && (data->num_msgs > 1) &&
```
Comment (cfriedt) [inline drivers/i2c/i2c_sam0.c:146]: this should be a bool

**Comment by cfriedt** [comment]
It's not immediately obvious - what necessitated `foo = value` => `data->foo = value`?

**Comment by rustypig91** [comment]
> It's not immediately obvious - what necessitated all of the changes from `foo` to `data->foo`, and how are the `foo` local variables still being used?

data->msgs and data->num_msgs are now shared variables with the isr. The isr will update these variables if it directly continues on with the next message.

File: drivers/i2c/i2c_sam0.c [hunk 2]
```diff
@@ -137,6 +139,17 @@ static void i2c_sam0_isr(const struct device *dev)
...
 
+	/*
+	 * Directly send/receive next message if it is in the same direction and
+	 * the current message has no stop or restart flags.
+	 */
+	const bool continue_next = (data->msg.size == 1) && (data->num_msgs > 1) &&
+				   ((data->msgs[0].flags & I2C_MSG_RW_MASK) ==
+				    (data->msgs[1].flags & I2C_MSG_RW_MASK)) &&
+				   !(data->msgs->flags & I2C_MSG_STOP) &&
+				   !(data->msgs->flags & I2C_MSG_RESTART) &&
```
Comment (anangl) [inline drivers/i2c/i2c_sam0.c:150]: The restart flag should be checked on the next message, not the current one.

**Comment by cfriedt** [comment]
> > It's not immediately obvious - what necessitated all of the changes from `foo` to `data->foo`, and how are the `foo` local variables still being used?
> 
> data->msgs and data->num_msgs are now shared variables with the isr. The isr will update these variables if it directly continues on with the next message.
> 
> This could be changed so that the isr instead only receives pointers to the old variables for less code change.

Hmm... Zephyr-style ISR handlers do have a data pointer but presumably that would used by the driver instance. I see no reason to block, but it would be great if you could look into that later @rustypig91.

**Review by cfriedt — APPROVED** [review]

File: drivers/i2c/i2c_sam0.c (hunk 2 above)
Comment (rustypig91) [reply drivers/i2c/i2c_sam0.c]: Fixed, also updated line 149 to make it more readable.

File: drivers/i2c/i2c_sam0.c [hunk 3]
```diff
@@ -137,6 +139,17 @@ static void i2c_sam0_isr(const struct device *dev)
 		return;
 	}
 
+	/*
+	 * Directly send/receive next message if it is in the same direction and
+	 * the current message has no stop or restart flags.
```
Comment (anangl) [inline drivers/i2c/i2c_sam0.c:144]: This comment became inaccurate after the recent change regarding the restart flag. It would be nice to update it.

**Review by anangl — APPROVED** [review]

**Comment by bryan-hunt** [comment]
How was this tested? This change causes every byte to get NACK'd if num_msgs  == 1

**Comment by rustypig91** [comment]
> How was this tested? This change causes every byte read to get NACK'd if num_msgs == 1

Tested against an ssd1306 so unfortunately limited testing on read messages :/
#38881 should fix this issue (it works when talking between two Arduino boards).

**Pull request merged**